src/tree/display.rs

```diff
@@ -1,12 +1,86 @@
-use crate::Node;
+use crate::{Node, Operator};
 use std::fmt::{Display, Error, Formatter};
 
+fn write_binary(f: &mut Formatter, node: &Node, op: &str) -> Result<(), Error> {
+    write!(
+        f,
+        "({} {} {})",
+        node.children.get(0).ok_or(Error)?,
+        op,
+        node.children.get(1).ok_or(Error)?
+    )
+}
+
+fn write_unary(f: &mut Formatter, node: &Node, op: &str) -> Result<(), Error> {
+    write!(f, "{}{}", op, node.children.get(0).ok_or(Error)?,)
+}
+
+fn write_root(f: &mut Formatter, node: &Node) -> Result<(), Error> {
+    for c in node.children.iter() {
+        write!(f, "{}", c)?
+    }
+    Ok(())
+}
+
+fn write_sequence(f: &mut Formatter, node: &Node, sep: &str) -> Result<(), Error> {
+    write!(f, "(")?;
+    for (i, c) in node.children.iter().enumerate() {
+        write!(f, "{}", c)?;
+        if i + 1 < node.children.len() {
+            write!(f, "{} ", sep)?;
+        }
+    }
+    write!(f, ")")
+}
+
 impl Display for Node {
     fn fmt(&self, f: &mut Formatter) -> Result<(), Error> {
-        self.operator.fmt(f)?;
-        for child in self.children() {
-            write!(f, " {}", child)?;
+        match &self.operator {
+            Operator::RootNode => write_root(f, self),
+            Operator::Add => write_binary(f, self, "+"),
+            Operator::Sub => write_binary(f, self, "-"),
+            Operator::Neg => write_unary(f, self, "-"),
+            Operator::Mul => write_binary(f, self, "*"),
+            Operator::Div => write_binary(f, self, "/"),
+            Operator::Mod => write_binary(f, self, "%"),
+            Operator::Exp => write_binary(f, self, "^"),
+            Operator::Eq => write_binary(f, self, "=="),
+            Operator::Neq => write_binary(f, self, "!="),
+            Operator::Gt => write_binary(f, self, ">"),
+            Operator::Lt => write_binary(f, self, "<"),
+            Operator::Geq => write_binary(f, self, ">="),
+            Operator::Leq => write_binary(f, self, "<="),
+            Operator::And => write_binary(f, self, "&&"),
+            Operator::Or => write_binary(f, self, "||"),
+            Operator::Not => write_unary(f, self, "!"),
+            Operator::Assign => write_binary(f, self, "="),
+            Operator::AddAssign => write_binary(f, self, "+="),
+            Operator::SubAssign => write_binary(f, self, "-="),
+            Operator::MulAssign => write_binary(f, self, "*="),
+            Operator::DivAssign => write_binary(f, self, "/="),
+            Operator::ModAssign => write_binary(f, self, "%="),
+            Operator::ExpAssign => write_binary(f, self, "^="),
+            Operator::AndAssign => write_binary(f, self, "&&="),
+            Operator::OrAssign => write_binary(f, self, "||="),
+            Operator::Tuple => write_sequence(f, self, ","),
+            Operator::Chain => write_sequence(f, self, ";"),
```

Comment:
Chain does not need to be surrounded by parentheses.

Reply:
I think it actually does, since (iiuc) it can occur as an argument to a function:

`f (g 4; 5)` shouldn't get turned into `f g 4; 5`

Reply:
Hm, I thought that chaining cannot occur within parentheses. This is not actually intended. But if it works that way at the moment, then that is how it is. I'll add some tests about this to the main branch to figure it out.

As long as the parentheses are required to make your planned tests work, they should stay.

Reply:
Ah I see you added these as tests already. Thanks! In that case, I agree with parentheses staying.